### ISSUE TITLE
[Feature]: Replace rule-based scoring with TF-IDF + Cosine Similarity for smarter project recommendations

### ISSUE THREAD

**Issue by Dhruvpandey1476** [komalharshita/DevPath#643]
## Problem Statement

The current recommendation engine in `utils/recommender.py` uses a hardcoded
rule-based scoring system with fixed integer weights:

| Match Type     | Score |
|----------------|-------|
| Skill match    | +3    |
| Level match    | +2    |
| Interest match | +2    |
| Time match     | +1    |

This approach has critical limitations:

- **No semantic matching** — "JS" will never match "JavaScript", "ML" won't match
  "Machine Learning", "py" won't match "Python"
- **Binary scoring** — a project is either a full match or zero; no partial credit
  for adjacent skills
- **Ignores project descriptions** — the `description`, `features`, and `tech_stack`
  fields in `projects.json` are never used for scoring, wasting rich textual signal
- **Scales poorly** — as more projects are added to the dataset, exact-match
  scoring becomes increasingly unreliable

## Proposed Solution

Replace or augment the scoring engine with a **TF-IDF + Cosine Similarity** pipeline
using `scikit-learn`, which is already a standard Python ML library and adds no
heavy dependencies.

### How it would work

1. At startup, build a **corpus** by concatenating each project's skills, interest,
   level, description, and tech_stack into a single text document per project
2. Fit a `TfidfVectorizer` on this corpus (can be cached in memory — no DB needed)
3. At request time, build a **query vector** from the user's selected skills,
   level, interest, and time
4. Compute **cosine similarity** between the query vector and all project vectors
5. Return the top-N projects ranked by similarity score

### Example Implementation Sketch

```python
# utils/recommender.py (new approach)
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def build_corpus(projects):
    docs = []
    for p in projects:
        text = " ".join([
            " ".join(p.get("skills", [])),
            p.get("level", ""),
            p.get("interest", ""),
            p.get("time", ""),
            p.get("description", ""),
            " ".join(p.get("tech_stack", [])),
        ])
        docs.append(text.lower())
    return docs

def recommend_projects(user_input, projects, top_n=3):
    corpus = build_corpus(projects)
    query = " ".join([
        " ".join(user_input.get("skills", [])),
        user_input.get("level", ""),
        user_input.get("interest", ""),
        user_input.get("time", ""),
    ]).lower()

    vectorizer = TfidfVectorizer()
    tfidf_matrix = vectorizer.fit_transform(corpus + [query])

    project_vectors = tfidf_matrix[:-1]
    query_vector = tfidf_matrix[-1]

    scores = cosine_similarity(query_vector, project_vectors).flatten()
    top_indices = np.argsort(scores)[::-1][:top_n]

    return [projects[i] for i in top_indices]
```

## Benefits Over Current Approach

| Feature                         | Current (Rule-based) | Proposed (TF-IDF) |
|---------------------------------|----------------------|--------------------|
| Handles "JS" → "JavaScript"     | ❌                   | ✅ (after stemming)|
| Uses project descriptions       | ❌                   | ✅                 |
| Partial skill overlap scoring   | ❌                   | ✅                 |
| Scales with dataset growth      | ❌                   | ✅                 |
| Zero new infrastructure needed  | ✅                   | ✅                 |
| Backward compatible API         | ✅                   | ✅                 |

## Dependencies

Only `scikit-learn` needs to be added to `requirements.txt`.
No database, no model training, no external API required.

## Scope

- Modify `utils/recommender.py`
- Add `scikit-learn` to `requirements.txt`
- Update/add tests in `tests/test_basic.py` to cover the new scorer
- Optionally expose the similarity score as a `match_score` field in the
  `/api/recommend` JSON response so the frontend can display a "% match"

## Difficulty Level

🟡 Intermediate / Advanced

## Labels

`enhancement`, `advanced`, `gssoc-26`

I would like to work on this under GSSoC'26

**Comment by KrutagyaKaneria** [comment]
Hi @komalharshita  👋

I would like to work on this enhancement under GSSOC ’26.

I carefully reviewed the current recommendation system and fully understand the limitations of the existing rule-based scoring approach. Replacing or augmenting it with a TF-IDF + Cosine Similarity pipeline would significantly improve recommendation quality, scalability, and semantic matching without requiring additional infrastructure.

I have experience working with Python, recommendation systems, scikit-learn, similarity scoring pipelines, and backend optimization, and I’d love to contribute this enhancement.

My implementation plan includes:

* Refactoring `utils/recommender.py` to support TF-IDF vectorization and cosine similarity ranking
* Building a reusable project corpus using:

  * skills
  * interest
  * level
  * time
  * description
  * tech_stack
* Generating dynamic query vectors from user preferences
* Returning top-N recommendations ranked by similarity score
* Preserving backward compatibility with the existing API structure
* Adding `match_score` support in API responses if appropriate
* Updating `requirements.txt` with `scikit-learn`
* Writing/expanding tests in `tests/test_basic.py` to validate recommendation quality and prevent regressions

I also understand the importance of keeping the implementation lightweight, fully in-memory, and efficient without introducing external APIs or database dependencies.

This enhancement would greatly improve the intelligence and usability of the recommendation engine as the project dataset grows, and I’d be excited to contribute it properly.

Please assign this issue to me 🚀

**Comment by Tamanna030507** [comment]
Hi @komalharshita,

I would like to work on this enhancement under GSSoC'26.

I have experience working with Python, ML workflows and recommendation systems and I would love to contribute by improving the current rule-based recommendation engine using TF-IDF and cosine similarity.

I can implement the vectorized recommendation pipeline, maintain backward compatibility with the existing API structure, optimize similarity scoring and add/update relevant test cases to ensure reliable recommendation quality.

Please assign this issue to me. Thank you!

**Comment by prathamsingh404** [comment]
Hi @komalharshita! 👋 I would love to tackle this enhancement as part of GSSoC '26.

The transition from a rigid, binary rule-based system to a semantic matching engine is a fantastic architectural upgrade. I'm very interested in applying Machine Learning techniques to solve practical backend problems, and building out this TF-IDF and cosine similarity pipeline aligns perfectly with my experience in Python and ML workflows. 

Here is my concrete plan to implement this smoothly, ensuring we keep the solution lightweight and fully in-memory:

1. **Dependency Management:** Add `scikit-learn` to `requirements.txt`.
2. **Corpus Generation:** Refactor `utils/recommender.py` to build a rich document for each project by concatenating `skills`, `interest`, `level`, `time`, `description`, and `tech_stack`.
3. **Vectorization & Scoring Engine:** Implement `TfidfVectorizer` to fit the project corpus, vectorize the incoming user preferences, and calculate the `cosine_similarity` to rank the top-N projects.
4. **API Enhancement:** Ensure backward compatibility with the existing API structure while exposing the new similarity score as a `match_score` field in the JSON response.
5. **Validation:** Expand the test suite in `tests/test_basic.py` to validate semantic matches (e.g., ensuring "JS" matches "JavaScript" via stemming/tokenization) and guarantee no regressions in performance.

I'm ready to start building this out and would ensure clean, well-documented code. Please let me know if I can be assigned to this issue! 🚀

**Comment by MJ606782** [comment]
.take

Hey @komalharshita ! This is a fantastic enhancement. Upgrading from a rigid exact-match scoring system to a TF-IDF + Cosine Similarity vector space model will dramatically elevate the relevance of the recommendations by utilizing rich text signals from project descriptions and handling partial token overlaps natively.

I would love to be assigned to this issue for GSSoC'26. Here is my proposed execution and optimization plan:

### 1. Vectorizer Optimization & Efficiency
* **Fix Corpus Leakage / Re-fitting:** In the basic sketch, `vectorizer.fit_transform(corpus + [query])` fits the vocabulary on the query at request time. I will optimize this by fitting the `TfidfVectorizer` **once** on the project corpus at startup (or caching it) and then using `.transform([query])` at request time. This avoids re-calculating the entire matrix for every API call.
* **Stopwords & Text Preprocessing:** I will configure the vectorizer with `stop_words='english'` and custom token splitting to clean up punctuation, ensuring short terms like "JS" and "ML" map cleanly without noise.

### 2. Implementation & Schema Updates (`utils/recommender.py`)
* Implement the unified corpus building blocks using robust `.get()` fallbacks.
* Calculate the similarity arrays using `scikit-learn`'s native spatial metrics and extract the indices using fast `numpy` array slicing (`np.argsort`).
* Append a normalized `match_score` float or percentage string to the returned payload so the frontend can dynamically render a match gauge.

### 3. Dependency Management & Testing
* Append `scikit-learn` and `numpy` to `requirements.txt`.
* Update `tests/test_basic.py` with mock project corpora to test exact matches, partial keyword overlaps (e.g., matching "Python" against a project description containing "py"), and verification that the `match_score` scales logically between `0.0` and `1.0`.

I'm ready to fork the repository and implement a highly optimized, clean pipeline for this. Please assign this to me!
